ffig/filters/capi_filter.py

```diff
@@ -398,3 +398,98 @@ def to_dotnet_return_value(t, rv):
     raise Exception(
         'Type {} has no defined dotnet return value translation (adding one may be trivial)'.format(
             t.name))
+
+
+# Java conversions
+def to_java_c_param(arg):
+    t = arg.type
+    if t.kind == TypeKind.INT:
+        return "{} {}".format(arg.type, arg.name)
+    if t.kind == TypeKind.DOUBLE:
+        return "{} {}".format(arg.type, arg.name)
+    if t.kind == TypeKind.POINTER:
+        if t.pointee.kind == TypeKind.RECORD:
+            return "Pointer {}".format(arg.name)
+    raise Exception(
+        'Type {} has no defined java C parameter translation (adding one may be trivial)'.format(
+            arg.type.name))
+
+
+def to_java_param(arg):
+    t = arg.type
+    if t.kind == TypeKind.INT:
+        return "{} {}".format(arg.type, arg.name)
+    if t.kind == TypeKind.DOUBLE:
+        return "{} {}".format(arg.type, arg.name)
+    if t.kind == TypeKind.POINTER:
+        if t.pointee.kind == TypeKind.RECORD:
+            return '{} {}'.format(
+                t.pointee.name.replace(
+                    'const ', ''), arg.name)
+    raise Exception(
+        'Type {} has no defined java parameter translation (adding one may be trivial)'.format(
+            arg.type.name))
+
+
+def to_java_output_param(t):
+    if t.kind == TypeKind.INT:
+        return "IntByReference"
+    if t.kind == TypeKind.DOUBLE:
+        return "DoubleByReference"
+    if t.kind == TypeKind.POINTER:
+        return "PointerByReference"
+    raise Exception(
+        'Type {} has no defined java output parameter translation (adding one may be trivial)'.format(
+            t.name))
+
+
+def to_java_output_value(t, rv):
+    if t.kind == TypeKind.INT:
+        return "IntByReference {} = new IntByReference();".format(rv)
+    if t.kind == TypeKind.DOUBLE:
+        return "DoubleByReference {} = new DoubleByReference();".format(rv)
+    if t.kind == TypeKind.POINTER:
+        return "PointerByReference {} = new PointerByReference();".format(rv)
```

Comment:
I'd have expected the template to have something like

```
{{returntype | to_java}} rv = new {{returntype | to_java}}();
```

is there a reason this is part of the text in the filter instead?

Reply:
Only consistency with the .cs template. I'm not sure why the dotnet template put the work in the filter rather than the template.

We'll need a revise-templates and filters pass soon to standardise names and approaches.

Did you have a strong preference which way we do things?